### ISSUE TITLE
Support iframes

### ISSUE THREAD

**Issue by giuseppeg** [reakit/reakit#456]
### Motivation

Hello there, I am evaluating Reakit to figure out if we can use it at work and one constraint we have is that we render our React application inside of an iframe and that when app code uses globals like `document`, these should be the iframe's. 

### Example

This is one of the many examples:

https://github.com/reakit/reakit/blob/5f51e395835ae1d1c26fc97699d02197db6ba3eb/packages/reakit-utils/src/hasFocusWithin.ts#L1-L4

Here you are referencing the global document. The fix here is easy enough:

```js
export function hasFocusWithin(element: Element) {
  const activeElement = element.ownerDocument.activeElement
  if (!activeElement) return false;
  return element.contains(activeElement);
}
```

### Possible implementations

Basically every time you reference `document` or `window` etc you'd need it to get it from an element (ref?) within the frame. Probably another option is to provide those globals via context.

Thanks in advance for your time, and thank you for the library it looks promising!

**Comment by diegohaz** [comment]
I'm not sure how it would be implemented using context? Do you have any CodeSandbox example of that?

**Comment by giuseppeg** [comment]
Using context is probably the last option I would look at but I'd provide `document` via a top level provider. I would try to avoid this as much as possible though and try to patch the library like in the example above using `ownerDocument`. 

Do you have a rough idea about how much it would take to implement this? I could spend half a day at work sometime.

**Comment by diegohaz** [comment]
The only problem I see is that there are several parts of the code using `document` and `window` without any reference to an element.

I still don't know if I understood it correctly. Are you saying that the value of `window` and `document` when being accessed from within an iframe are actually referencing the parent window? I'm not experienced with iframes, but I'd expect that it worked the same way from within the iframe, and if you wanted to access the parent window you would have to use `window.parent` or something similar.

**Comment by giuseppeg** [comment]
> The only problem I see is that there are several parts of the code using document and window without any reference to an element.

That is an issue indeed! Maybe functions and components could have arguments or props which default to globals?

> I still don't know if I understood it correctly. Are you saying that the value of window and document when being accessed from within an iframe are actually referencing the parent window?

Correct, imagine that the library code runs in the the host page (parent window) and creates an `iframe` and renders the React application inside of the iframe. 

When you access `document`, `window` or any other global those would be the parent's (host) window.

This is a problem when you want to find out what is the `activeElement`, when you want to position a popover or use portals to append to `document.body`. To give you a practical example, at work we build a PDF Viewer that can be embedded in a page https://web-examples.pspdfkit.com and we render inside of iframe to isolate styles and other stuff.

**Comment by diegohaz** [comment]
The code lives in the parent window. Got it!

I think the first step is to map all the usages of `window` and `document` throughout the codebase and determine their situation, for example:
- works
- doesn't work
- may use an element from the scope
- may receive an element from parameters
- may receive an element from the context
- can't receive element (e.g. [`document.body` being used on the top level](https://github.com/reakit/reakit/blob/208bcb65a749e2a6854e7102f6be218ac8922558/packages/reakit/src/Portal/Portal.tsx#L12))

**Comment by diegohaz** [comment]
Some progress has been made on this. There are still these lines to update:

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Dialog/__utils/useFocusTrap.ts#L59

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Disclosure/__utils/useWarningIfMultiple.ts#L12

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Form/FormPushButton.ts#L61

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Form/FormRemoveButton.ts#L56

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Form/FormRemoveButton.ts#L70

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Form/FormRemoveButton.ts#L78

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Form/__utils/getFirstInvalidInput.ts#L3

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Portal/Portal.tsx#L13

https://github.com/reakit/reakit/blob/49db6c007ea82fddf959448328b5f5d572defa40/packages/reakit/src/Portal/Portal.tsx#L27

**Comment by giuseppeg** [comment]
Thank you a lot @diegohaz!

**Comment by diegohaz** [comment]
Only `Portal` left now! I'm not sure what to do with that as `document` is assigned at the module's top-level scope.

Maybe just let people use `PortalContext.Provider` and pass their own container element.

**Comment by giuseppeg** [comment]
yeah that'd work. Here's how I do that at work https://github.com/giuseppeg/react-layers-manager/blob/master/src/LayersManager.js (wrapping the app also allows to stack things properly)
